---
title: "Figures and tables: _E. coli_ genomics"
author: "Joe Lewis"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggtree)
library(ape)
library(phytools)
library(tidyverse)
library(blantyreESBL)
library(lubridate)
library(ggnewscale)
library(countrycode)
library(viridis)
library(blantyreESBL)
library(kableExtra)
library(patchwork)
library(here)
library(ggnewscale)
library(scales)
library(ggstance)
```

## Introduction

This document generates figures and tables for Joe's _E. coli_ genomics paper.

### Assembly stats and accession numbers

```{r accession-table}

btESBL_sequence_sample_metadata %>%
  filter(lane %in% btESBL_coregene_tree_esco$tip.label) %>%
  left_join(
    rbind(
      read_tsv(
        "~/Documents/PhD/Thesis/bookdown/chapter_7/checkm_quast/D1/transposed_report.tsv"
      ) ,
      read_tsv(
        "~/Documents/PhD/Thesis/bookdown/chapter_7/checkm_quast/D220190318/transposed_report.tsv"
      ),
      read_tsv(
        "~/Documents/PhD/Thesis/bookdown/chapter_7/checkm_quast/D220190503/transposed_report.tsv"
      )
    ) %>%
      transmute(
        lane = gsub("\\.contigs_spades", "",  Assembly),
        number_of_contigs = `# contigs`,
        N50 = N50
      ),
    by = "lane"
  ) %>% 
  rename(date_of_collection = data_date) %>% 
  select(accession, lane, supplier_name, pid, date_of_collection, number_of_contigs, N50) -> t1

t1 %>% 
  kbl(caption = "Accession numbers and assembly statistics for included samples") %>% 
  kable_classic(full_width = FALSE)

write_csv(t1, 
          here("manuscript/esco/supp_tables/SUPP_T1_accession_numbers_and_metadata.csv"))

```

### Phylogroup and MLST

```{r mlst-and-phylogroup, fig.height = 4, fig.width = 7, fig.cap = "FIGURE 1: Sequency types (A) and phylogroups (B) of included isolates"}

left_join(
  read_csv("~/Documents/PhD/Thesis/bookdown/chapter_7/phylogroup_and_mlst/mlst.csv"),
  read_csv("~/Documents/PhD/Thesis/bookdown/chapter_7/phylogroup_and_mlst/phylogroups.csv"),
  by = c("lane" = "Lane")
  ) -> 
  pgroup_mlst

pgroup_mlst %>% 
  group_by(ST) %>% 
  mutate(n = n()) %>% 
  filter(n > 2) %>% 
  ggplot(aes(fct_rev(fct_infreq(ST)), fill = Phylogroup)) +
  geom_bar() +
  coord_flip() +
  theme_bw() +
  theme(legend.position = "none") +
  labs(x = "ST", y = "n") -> p1 
#  scale_fill_viridis_d(option = "cividis") -> p1
  
pgroup_mlst %>% 
  ggplot(aes(fct_infreq(Phylogroup), fill = Phylogroup)) +
  geom_bar() +
  theme_bw() +
  theme(legend.position = "none", 
        axis.text.x = element_text(angle = 45,
                                   hjust = 1)) +
#  scale_fill_viridis_d(option = "cividis") +
  labs(x = "Phylogroup", y = "n") -> p2
  
(p1 | p2) + plot_annotation(tag_levels = "A") -> mlst_plot

ggsave(here("manuscript/esco/figures/FIG_1_mlstplot.svg"), mlst_plot,
       width = 7, height = 4)

mlst_plot 
```

### Malawi phylogeny

```{r malawi-tree, fig.height=9, fig.width= 9, fig.cap = "SUPPLEMENTARY FIGURE 1: Midpoint-rooted maximum-likelohood phylogeny of study isolates shwoing phylogroup and multilocus sequence type."}

pgroup_mlst %>% 
  as.data.frame() -> pgroup_mlst

rownames(pgroup_mlst) <- pgroup_mlst$lane

pgroup_mlst %>% 
  select(lane, ST) %>% 
  group_by(ST) %>% 
  mutate(n = n(),
         ST = if_else(n ==1, "Other", ST)) %>% 

  mutate(ST = if_else(!ST %in% c("Novel", "Other"), 
                      paste0("ST", ST), ST)) %>% 
  arrange(fct_infreq(ST)) %>% 
  pivot_wider(id_cols = lane, 
              names_from = ST,
              values_from = ST,
              values_fn = length,
              values_fill = 0) %>% 
  mutate(across(everything(), as.character)
         ) %>% 
  relocate(Other, .after = everything()) %>% 
  as.data.frame() ->
  mlst_onehot

rownames(mlst_onehot) <- mlst_onehot$lane

(
  ggtree(btESBL_coregene_tree_esco) %>%
    gheatmap(
      select(pgroup_mlst, Phylogroup),
      width = 0.1,
      color = NA,
      font.size = 4,
      colnames_angle = 90,
      colnames_position = "top",
      colnames_offset_y = 3,
      hjust = 0
    ) +
    scale_fill_manual(
      values = hue_pal()(8),
      name = "Phylogroup") + 
   new_scale_fill()
) %>%
  gheatmap(select(mlst_onehot, -lane),
           font.size = 2.5,
           color = NA,
           colnames_angle = 90,
           offset = .03,
           colnames_offset_y = 3,
           colnames_position = "top",
           hjust = 0)  +
  scale_fill_manual(values = c("lightgrey", "black"), guide = FALSE) +
  ylim(NA, 560)  + 
  annotate("text", x= 0.28, y =520, label = "Sequence Type") -> malawi_treeplot

 malawi_treeplot
 
ggsave(here("manuscript/esco/figures/SUPP_FIG_1_malawi_treeplot.svg"), malawi_treeplot, width = 9, height = 9)


```

### AMR determinents

```{r amrplot, fig.width= 6, fig.height=10}

read_csv("~/Documents/PhD/Thesis/bookdown/chapter_7/amr/srst2_pres_abs.csv") %>% filter(name %in% btESBL_coregene_tree_esco$tip.label) -> amr.ariba
  

amr.qrdr <- read_csv( "~/Documents/PhD/Thesis/bookdown/chapter_7/amr/qrdr_mut_presence.csv")

# get beta lactamase phenotype and tidy -------------------------------------

bls <- read_tsv(
  "https://ftp.ncbi.nlm.nih.gov/pathogen/betalactamases/Allele.tab"
  )


bls %>% 
  rename_with(~ tolower(gsub(" ", "_", .x))) %>% 
  rename_with(~ gsub("#", "", .x)) %>% 
  mutate(allele_name = gsub("-", "_", allele_name),
         class = case_when(
           grepl("carbapenem-hydrolyzing", 
                 curated_gene_product_name) ~ "Carbapenemase",
           grepl("metallo-beta-lactamase", 
                 curated_gene_product_name) ~ "Carbapenemase",
           grepl("extended-spectrum beta-lactamase", 
                 curated_gene_product_name) ~ "ESBL",
           grepl("class C", 
                 curated_gene_product_name) ~ "AmpC",
           grepl("beta-lactamase", 
                 curated_gene_product_name) ~ "Penicillinase"
         )) -> 
  bls




# add class ro which resistnce is conferred --------------------------------

quinolone <- "Par|Gyr|Par|Qnr|Qep|Nor|GyrA|GyrB|ParC|ParE"
tetracycline<- "Tet"
sulphonamide <- "Sul"
aminoglycoside <- "Str|Aad|Aac|Aph|Rmt|APH"
streptothricin <- "Sat"
macrolide <- "Mph|Mdf|Erm|Ere"
fosfomycin <- "Fos"
chloramphenicol <- "Cat|FloR|Cml"
trimethoprim <- "Dfr"
rifampicin <- "Arr"
ESBL <- "SHV_12"
penicillinase <- "OKP|SCO|LEN|LAP"
ampc <- "CMY"

amr.ariba %>% 
  pivot_longer(-name, names_to = "gene") %>% 
  bind_rows(
    amr.qrdr %>% 
    pivot_longer(-sample, names_to = "gene") %>% 
      rename(name = sample)
    ) %>% 
  filter(gene != "AmpH") %>% 
  left_join(select(bls, allele_name, class),
            by = c("gene" = "allele_name")) %>% 
  mutate(class = case_when(
    str_detect(gene, quinolone) ~ "Quinolone",
    str_detect(gene, tetracycline) ~ "Tetracycline",
    str_detect(gene, sulphonamide) ~ "Sulphonamide",
    str_detect(gene, aminoglycoside) ~ "Aminoglycoside",
    str_detect(gene, streptothricin) ~ "Streptothricin",
    str_detect(gene, macrolide) ~ "Macrolide",
    str_detect(gene, fosfomycin) ~ "Fosfomycin",
    str_detect(gene, chloramphenicol) ~ "Chloramphenicol",
    str_detect(gene, rifampicin) ~ "Rifampicin",
    str_detect(gene,trimethoprim) ~ "Trimethoprim",
    str_detect(gene,ESBL) ~ "ESBL",
    str_detect(gene,penicillinase) ~ "Penicillinase",
    str_detect(gene,ampc) ~ "AmpC",
    TRUE ~ class
  )) ->
  amr.ariba.long

### plot prevalence ----------------------------------------


amr.ariba.long %>% 
  filter(value == 1) %>% 
  filter(!is.na(class)) %>% 
  group_by(class) %>% 
  mutate(n_class = length(class),
         n_genes_in_class = n_distinct(gene)) %>% 
  select(class, n_class,n_genes_in_class) %>% 
  unique() %>% 
  arrange(n_class) %>% 
  ungroup() %>% 
  mutate(end = cumsum(n_genes_in_class),
         start = lag(end, default = 0),
         textpos = start+0.6 + 0.5*(end-start)) -> annotate.df

amr.ariba.long %>% 
  filter(value == 1) %>% 
  filter(!is.na(class)) %>% 
  group_by(class) %>% 
  mutate(n_class = length(class),
         n_genes_in_class = n_distinct(gene),
         gene = gsub("_", "-", gene)) %>% 
ggplot( aes(fct_reorder(fct_rev(fct_infreq(gene)), n_class), 
            fill = class)) + 
  geom_bar() +
  theme_bw() +
  coord_flip(ylim = c(-5,450),clip = "off", expand = FALSE, xlim = c(0, 76)) +
  annotate(geom = "segment",
             x = annotate.df$start +0.5+0.2, 
             xend = annotate.df$end +0.5-0.2, 
             y = -115, yend = -115 ) +
  annotate(geom = "text", y = -125,
           x = annotate.df$textpos,
           label = annotate.df$class, 
           size = 3, 
           hjust = 1) +
  labs(y = "Number") + 
  theme(plot.margin = unit(c(0.2,0.5,0.2,4), "cm"),
        axis.title.y  = element_blank(),
        legend.position = "none") -> amrplot

amrplot
ggsave(filename = here("manuscript/esco/figures/FIG_2_amr_plot.svg"),plot = amrplot,
       width =6,
       height = 10
)

```

### Global phylogeny

```{r global_phylogeny, fig.height= 10, fig.width = 10}


tree <- read.tree(here("data_processed/esco/global_tree/all_ESCO_core_snps.aln.treefile"))

midpoint.root(tree) -> tree


#### get metadata--------------------------------

metadata <- read_csv(
  here("data_processed/esco/global_tree/all_ESCO_metadata.csv"))

accession_global <- read_csv(
  here("data_raw/e_coli/global_accession_numbers.csv"))

metadata %>% 
  mutate(
    Lane.Name = gsub("#", "_", Lane.Name ),
    Sample = gsub("_aEPEC", "", Sample),
    Sample = gsub("_","", Sample),
    Sample = gsub("CFn", "", Sample),
    Sample = gsub("sec","", Sample),
    Sample = gsub("^G|R","", Sample),
    Sample = gsub("-1|-1A|-2|-2A|-2B|-3A|-3B$","", Sample)) %>% 
  left_join(
  bind_rows(
  read_csv(
    "/Users/joelewis/Documents/Sanger/context_genomes/mentzer_metadata_table_blanks_rem.csv"
  ) %>%
    filter(!is.na(Designation)) %>%
    transmute(Sample = Designation,
              Year = isolation) %>% 
    left_join(
      read_csv(here("data_raw/e_coli/ilbnames.csv")) %>%  # coreect some names using
                                                   # sanger pf data and 
                                                   # accession numbers
        transmute(Sample = `Sample Name`, 
                  update_name = gsub("_", "", `Supplier Name`)),
      by = "Sample") %>% 
    mutate(Sample = if_else(!is.na(update_name), update_name, Sample)) %>% 
    select(-update_name),
  read_csv(
    "/Users/joelewis/Documents/Sanger/context_genomes/ingle_gems_metadata.csv"
  ) %>%
    transmute(Sample = Isolate,
              Year = `Year of Isolation`)
  ),
  by = "Sample") %>% 
  left_join(btESBL_sequence_sample_metadata %>% 
              transmute(Lane.Name = lane,
                        dassim_year = year(data_date)),
            by = "Lane.Name") %>% 
  left_join(
    read_csv(here("data_raw/e_coli/musicha_Ecoli_MetaData.csv")) %>% 
      transmute(Lane.Name = gsub("#", "_", Lane),
                musicha_year = Year),
    by = "Lane.Name") %>% 
  mutate(Year = as.numeric(
    if_else(Year == "unknown", NA_character_, Year)
  ),
    Year = case_when(
    grepl("DASSIM", study) ~ dassim_year,
    study == "runchaeron" ~ 2015,
    study == "musicha" ~ musicha_year,
    TRUE ~ Year)
  ) %>% 
  left_join(
    accession_global %>% 
      transmute(Lane.Name = gsub("#", "_", `Lane Name`),
                accession = `Lane Acc`),
    by = "Lane.Name"
  ) %>% 
  select(-c(dassim_year, musicha_year)) ->
           metadata

metadata %>% 
  transmute(
    accession = accession,
    lane = Lane.Name,
    source = Source,
    country = Country,
    source = source,
    year = Year,
    study = case_when(
      grepl("DASSIM", study) ~ "This study",
      study == "ingle_GEMS" ~ "Ingle",
      study == "mentzer" ~ "Mentzer",
      study == "musicha" ~ "Musicha",
      study == "runchaeron" ~ "Runchaeron")) %>%
  filter(study != "This study")  %>% 
  write_csv(
          here("manuscript/esco/supp_tables/SUPP_T2_global_accession_numbers_and_metadata.csv"))


metadata %>%
  mutate(
    Continent =
      countrycode(
        sourcevar = metadata$Country,
        origin = "country.name",
        destination = "un.region.name"
      ),
    Source = case_when(
      Source %in% c("Canal",
                    "Farm",
                    "Untreated hospital sewage") ~ "Environment",
      Source %in% c("Blood",
                    "CSF",
                    "Urine",
                    "Pus" ,
                    "Sputum")  ~ "Human - invasive",
      Source %in% c("stool",
                    "Stool" ,
                    "RS") ~ "Human - stool"
    ),
    this_study = if_else(grepl("DASSIM", study), "Yes", NA_character_),
    Country = case_when(
      Country == "Malawi" & grepl("DASSIM", study) ~ "Malawi: this study",
      Country == "Malawi" &
        !grepl("DASSIM", study) ~ "Malawi: previous study",
      TRUE ~ "Other"
    )
  ) %>%
  as.data.frame() ->
  metadata




rownames(metadata) <- metadata$Lane.Name

#### AMR --------------------------------

bind_rows(
  read.csv(
    here(
      "data_processed/esco/global_tree/srst2_summary_cleaned_pres_abs.csv"
    )
  ),
  read_csv(
    "~/Documents/PhD/Thesis/bookdown/chapter_7/amr/srst2_pres_abs.csv"
  )
) %>%
  mutate(across(everything(), ~ replace_na(.x, 0)))  %>%
  rename(sample = name) %>%
  pivot_longer(-sample) %>%
  filter(value == 1) %>%
  left_join(select(bls, allele_name, class),
            by = c("name" = "allele_name")) %>%
  mutate(class = case_when(
    grepl("CMY", name) ~ "ESBL/AmpC",
    class == "ESBL" ~ "ESBL/AmpC",
    TRUE ~ class
  )) %>%
  filter(class %in% c("ESBL/AmpC", "Carbapenemase") &
           value == 1) %>%
  pivot_wider(
    id_cols = sample,
    names_from = class,
    values_from = class,
    values_fn = length
  ) %>%
  mutate(
    `ESBL/AmpC` = if_else(!is.na(`ESBL/AmpC`), "ESBL/AmpC", NA_character_),
    Carbapenemase = if_else(!is.na(Carbapenemase), "Carbapenemase", NA_character_)
  ) %>%
  as.data.frame() -> amr_global

rownames(amr_global) <- amr_global$sample

# plot tree -------------------------------

(
  (
    (
      (
        (ggtree(tree) %<+% select(metadata, Lane.Name, Country))   %>%
          gheatmap(
            select(metadata, Continent),
            font.size = 4,
            width = 0.03,
            colnames_position = "top",
            color = NA,
            colnames_offset_y = 5,
            colnames_angle = 90,
            hjust = 0,
            offset = 0.006
          )  +
          scale_fill_manual(
            values = viridis_pal(option = "plasma")(5)[c(1, 2, 3)],
            na.translate = FALSE,
            name = "Continent"
          ) +
          new_scale_fill()
      ) %>%
        gheatmap(
          select(metadata, Country),
          font.size = 4,
          width = 0.03,
          colnames_position = "top",
          color = NA,
          colnames_offset_y = 5,
          colnames_angle = 90,
          hjust = 0,
          offset = 0.003
        ) +
        scale_fill_manual(values = c(
          viridis_pal(option = "cividis", )(4)[2:1],
          "white"
        ),
        name = "Country") +
        new_scale_fill()
    )
    %>%
      gheatmap(
        select(amr_global,-sample),
        font.size = 3,
        width = 0.06,
        colnames_position = "top",
        color = NA,
        colnames_offset_y = 5,
        colnames_angle = 90,
        offset = 0.012,
        hjust = 0
      ) + scale_fill_manual(
        values = viridis_pal(option = "magma")(5)[c(2, 3)],
        na.translate = FALSE,
        name = "Beta-\nlactamase",
        guide = FALSE
      ) +
      new_scale_fill()
  ) %>%
    gheatmap(
      select(metadata, `Year`),
      font.size = 4,
      width = 0.03,
      colnames_position = "top",
      color = NA,
      colnames_offset_y = 5,
      colnames_angle = 90,
      hjust = 0
    )  +
    scale_fill_viridis_c(name = "Year")  +
    new_scale_fill()
) %>%
  gheatmap(
    select(metadata, Source),
    font.size = 4,
    width = 0.03,
    colnames_position = "top",
    color = NA,
    colnames_offset_y = 5,
    colnames_angle = 90,
    hjust = 0,
    offset = 0.009
  ) + scale_fill_viridis_d(option = "magma", name = "Source") +
  geom_cladelabel(node = 1351, label = "ST131", fontsize = 2.5, offset = -0.0009,
                  offset.text = -0.0005) +
  geom_cladelabel(node = 1864, label = "ST410", fontsize = 2.5,
                  offset.text = -0.0005) +
  geom_cladelabel(node = 2020, label = "ST167", fontsize = 2.5,
                  offset.text = -0.0005) +
  geom_treescale(x = 0.04, y = 1200, offset = 10) +
  ylim(NA, 1500) -> global_tree_plot

ggsave(here("manuscript/esco/figures/FIG_3_global_tree_plot.svg"), 
       global_tree_plot, width = 8, height = 12)

ggsave(here("manuscript/esco/figures/FIG_3_global_tree_plot.pdf"), 
       global_tree_plot, width = 8, height = 12)


global_tree_plot
  

```

### ST167 tree

```{r st167 tree, fig.width= 10, fig.height=9}

st167_metadata <- read_tsv(here("data_raw/e_coli/st167.tsv"))



st167_metadata %>%
  select(
    Uberstrain,
    Name,
    `Data Source(Accession No.;Sequencing Platform;Sequencing Library;Insert Size;Experiment;Status)`,
    `Source Niche`,
    `Source Details`,
    Country,
    `Collection Year`,
    ST
  ) %>%  separate_rows(
    `Data Source(Accession No.;Sequencing Platform;Sequencing Library;Insert Size;Experiment;Status)`,
    sep = ","
  ) %>%
  separate(
    `Data Source(Accession No.;Sequencing Platform;Sequencing Library;Insert Size;Experiment;Status)`,
    into = c(
      "accession",
      "platform",
      "library",
      "insert_size",
      "experiment"
    ),
    sep = ";"
  ) ->
  st167_metadata


read.tree(here("data_processed/esco/st167/iqtree/clean.full.filtered_polymorphic_sites.ref_removed.snpsites.fasta.treefile")) -> st167_tree

midpoint.root(st167_tree) -> st167_tree

st167_tree$tip.label <- gsub("_filtered","", st167_tree$tip.label)

st167_metadata %>% 
  filter(platform == "ILLUMINA") %>% 
  filter(accession %in% st167_tree$tip.label) %>% 
  bind_rows(
    btESBL_sequence_sample_metadata %>% 
      filter(lane %in% st167_tree$tip.label) %>% 
      transmute(accession = lane,
                `Collection Year` = year(data_date),
                `Source Niche` = "Human",
                Country = "Malawi")
  ) %>% 
  rename(Year = `Collection Year`) %>% 
  as.data.frame() -> st167_metadata

st167_metadata %>% 
  filter(Country != "Malawi") %>% 
  transmute(accession = accession,
            source = `Source Niche`,
            country = Country,
            year = Year) %>% 
  write_csv(here("manuscript/esco/supp_tables/SUPP_T3_st167_accession_numbers_and_metadata.csv"))
            
            

st167_metadata$malawi <- st167_metadata$Country == "Malawi"

st167_metadata$Continent <- 
  countrycode(
    sourcevar = st167_metadata$Country,
    origin = "country.name",
    destination = "un.region.name"
  )

rownames(st167_metadata) <- st167_metadata$accession



amr.ariba <- read_csv(here("data_processed/esco/st167/ariba/st167_ariba_srst2_summary.csv"))

amr.ariba %>% 
  mutate(name = gsub("\\./", "", name),
         name = gsub("/report.tsv", "", name),
         name = gsub("_filtered", "", name),
         name = gsub("#","_", name)) %>% 
  pivot_longer(-name, 
               names_to= c( "cluster", ".value"), 
               names_sep = "\\.") %>% 
  mutate(gene = sapply(str_split(ref_seq, "__"), function(x) x[3])) %>% 
  filter(match == "yes") %>% 
  mutate(gene = case_when(
    gene == "TEM_95" ~ "TEM_1",
    TRUE ~ gene
  )) %>% 
  select(name, gene) %>% 
  pivot_wider(names_from = gene, 
              values_from = gene,
              values_fn = length,
              values_fill = 0) %>%
  mutate(across(everything(), as.character)) %>% 
  as.data.frame() ->
  amr.ariba

rownames(amr.ariba) <- amr.ariba$name



amr.ariba %>% 
  rename(sample = name) %>% 
  pivot_longer(-sample) %>% 
  filter(value == 1) %>% 
  left_join(
    select(bls, allele_name, class),
    by = c("name" = "allele_name")
  ) %>%  
  mutate(class = case_when(
    grepl("CMY", name) ~ "ESBL/AmpC",
    class == "ESBL" ~ "ESBL/AmpC",
    TRUE ~ class
        )) %>% 
  filter(class %in% c("ESBL/AmpC", "Carbapenemase") & value == 1) %>%
  mutate(name = gsub("_", "-", name)) %>% 
  arrange(class, fct_infreq(name)) %>% 
  pivot_wider(id_cols = "sample",
              names_from = name,
              values_from = class) %>% 
  as.data.frame() %>% 
  select(-`TEM-15`) ->    # not in df?
  esbl_cpe.ariba

rownames(esbl_cpe.ariba) <- esbl_cpe.ariba$sample

(
(ggtree(treeio::tree_subset (st167_tree, 320, levels_back = 0)) %<+% (select(st167_metadata, accession, Country) %>% mutate(Country = if_else(Country == "Malawi", "Malawi",NA_character_)))  %>%
    gheatmap(
      select(st167_metadata, `Year`),
      font.size = 4,
      width = 0.03,
      colnames_position = "top",
      color = NA,
      colnames_offset_y = 5,
      colnames_angle = 90,
      hjust = 0
    )  + 
    scale_fill_viridis(name = "Year") +
    new_scale_fill()) %>% 
  gheatmap(
    select(st167_metadata, Continent),
    font.size = 4,
    width = 0.03,
    colnames_position = "top",
    color = NA,
    colnames_offset_y = 5,
    colnames_angle = 90,
    offset = 0.00003,
    hjust = 0
  ) + 
  scale_fill_viridis_d(option = "plasma", name = "Continent") + 
  geom_tippoint(aes(color = Country), na.rm = TRUE, size = 0.5) + 
  scale_color_manual(values  = "red", na.translate = FALSE) +
    new_scale_fill() 
) %>% 
  gheatmap(
  select(esbl_cpe.ariba, -sample),
  font.size = 3,
  width = 0.5,
  colnames_position = "top",
  color = NA,
  colnames_offset_y = 5,
  colnames_angle = 90,
  offset = 0.00006,
  hjust = 0
) + scale_fill_manual(values = viridis_pal(option = "magma")(5)[c(2,3)],
                      na.translate = FALSE,
                      name = "Beta-\nlactamase") +
  ylim(NA, 350) +
  geom_treescale(x = 0.0002, y = 250, offset = 2) -> st167_globaltree_plot

st167_globaltree_plot

ggsave(here("manuscript/esco/figures/FIG_4_st167_treeplot.svg"),
       st167_globaltree_plot, width = 9.5, height = 10)
ggsave(here("manuscript/esco/figures/FIG_4_st167_treeplot.pdf"),
       st167_globaltree_plot, width = 9.5, height = 10)

```

### ST410 trees

```{r st410 tree, fig.height = 10, fig.width = 12}


st410_metadata <- read_tsv(here("data_raw/e_coli/st410.tsv"))



st410_metadata %>%
  select(
    Uberstrain,
    Name,
    `Data Source(Accession No.;Sequencing Platform;Sequencing Library;Insert Size;Experiment;Status)`,
    `Source Niche`,
    `Source Details`,
    Country,
    `Collection Year`,
    ST
  ) %>%  separate_rows(
    `Data Source(Accession No.;Sequencing Platform;Sequencing Library;Insert Size;Experiment;Status)`,
    sep = ","
  ) %>%
  separate(
    `Data Source(Accession No.;Sequencing Platform;Sequencing Library;Insert Size;Experiment;Status)`,
    into = c(
      "accession",
      "platform",
      "library",
      "insert_size",
      "experiment"
    ),
    sep = ";"
  ) ->
  st410_metadata



read.tree(here("data_processed/esco/st410/iqtree/clean_full.filtered_pollymorphic_sites.ref_removed.snpsites.fasta.treefile")) -> st410_tree

midpoint.root(st410_tree) -> st410_tree

#ggtree(treeio::tree_subset (st410_tree, 1083, levels_back = 0)) + geom_treescale()

st410_tree$tip.label <- gsub("_filtered","", st410_tree$tip.label)

st410_metadata %>% 
  filter(platform == "ILLUMINA") %>% 
  filter(accession %in% st410_tree$tip.label) %>% 
  bind_rows(
    btESBL_sequence_sample_metadata %>% 
      filter(lane %in% st410_tree$tip.label) %>% 
      transmute(accession = lane,
                `Collection Year` = year(data_date),
                `Source Niche` = "Human",
                Country = "Malawi")
  ) %>% 
  rename(
    Year = `Collection Year`) %>% 
  as.data.frame() -> st410_metadata

st410_metadata %>% 
  filter(Country != "Malawi") %>% 
  transmute(accession = accession,
            source = `Source Niche`,
            country = Country,
            year = Year) %>% 
  write_csv(here("manuscript/esco/supp_tables/SUPP_T4_st410_accession_numbers_and_metadata.csv"))

st410_metadata$malawi <- st410_metadata$Country == "Malawi"


st410_metadata$Continent <- 
  countrycode(
    sourcevar = st410_metadata$Country,
    origin = "country.name",
    destination = "un.region.name"
  )

rownames(st410_metadata) <- st410_metadata$accession

amr.ariba410 <- read_csv(here("data_processed/esco/st410/ariba/st410_ariba_srst2_summary.csv"))

amr.ariba410 %>% 
  mutate(name = gsub("\\./", "", name),
         name = gsub("/report.tsv", "", name),
         name = gsub("_filtered", "", name),
         name = gsub("#","_", name)) %>% 
  pivot_longer(-name, 
               names_to= c( "cluster", ".value"), 
               names_sep = "\\.") %>% 
  mutate(gene = sapply(str_split(ref_seq, "__"), function(x) x[3])) %>% 
  filter(match == "yes") %>% 
  mutate(gene = case_when(
    gene == "TEM_95" ~ "TEM_1",
    TRUE ~ gene
  )) %>% 
  select(name, gene) %>% 
  pivot_wider(names_from = gene, 
              values_from = gene,
              values_fn = length,
              values_fill = 0) %>%
  mutate(across(everything(), as.character)) %>% 
  as.data.frame() ->
  amr.ariba410

rownames(amr.ariba410) <- amr.ariba410$name




amr.ariba410 %>% 
  rename(sample = name) %>% 
  pivot_longer(-sample) %>% 
  filter(value == 1) %>% 
  left_join(
    select(bls, allele_name, class),
    by = c("name" = "allele_name")
  ) %>% mutate(class = case_when(
    grepl("CMY", name) ~ "ESBL/AmpC",
    class == "ESBL" ~ "ESBL/AmpC",
    TRUE ~ class
        )) %>% 
  filter(class %in% c("ESBL/AmpC", "Carbapenemase") & value == 1) %>%
  mutate(name = gsub("_", "-", name)) %>% 
  arrange(class, fct_infreq(name)) %>% 
  pivot_wider(id_cols = "sample",
              names_from = name,
              values_from = class) %>% 
  as.data.frame() -> 
#  select(-`TEM-15`) ->    # not in df?
  esbl_cpe.ariba410

rownames(esbl_cpe.ariba410) <- esbl_cpe.ariba410$sample

(  
  (ggtree(treeio::tree_subset (st410_tree, 1083, levels_back = 0)) %<+% (select(st410_metadata, accession, Country) %>% mutate(Country = if_else(Country == "Malawi", "Malawi",NA_character_)))  %>%
     gheatmap(
       select(st410_metadata, `Year`),
       font.size = 4,
       width = 0.03,
       colnames_position = "top",
       color = NA,
       colnames_offset_y = 5,
       colnames_angle = 90,
       hjust = 0
     )  + 
     scale_fill_viridis(name = "Year") +
     new_scale_fill()) %>% 
  gheatmap(
    select(st410_metadata, Continent),
    font.size = 4,
    width = 0.03,
    colnames_position = "top",
    color = NA,
    colnames_offset_y = 5,
    colnames_angle = 90,
    offset = 0.00015,
    hjust = 0
  ) + 
  scale_fill_viridis_d(option = "plasma", name = "Continent") + 
  geom_tippoint(aes(color = Country), na.rm = TRUE, size = 0.5) + 
  scale_color_manual(values  = "red", na.translate = FALSE) +
  new_scale_fill() 
) %>% 
  gheatmap(
    select(esbl_cpe.ariba410, -sample),
    font.size = 3,
    width = 0.5,
    colnames_position = "top",
    color = NA,
    colnames_offset_y = 5,
    colnames_angle = 90,
    offset = 0.0003,
    hjust = 0
  ) + scale_fill_manual(values = viridis_pal(option = "magma")(5)[c(2,3)],
                        na.translate = FALSE,
                        name = "Beta-\nlactamase") +
  ylim(NA, 600) +
  geom_treescale(x = 0.0002, y = 250, offset = 5) +
   geom_highlight(aes(node = 568) ,alpha = 0.3) +
  geom_cladelabel(
    node = 568,
    label = "B4/H24RxC",
    angle = 90,
    offset = -0.0006,
    offset.text = -0.00009,
    barsize = 0,
    hjust = 0.5,
    fontsize = 3,
    color = "steelblue"
  )  -> st410_globaltree_plot

ggsave(here("manuscript/esco/figures/FIG_5_st410_treeplot.svg"),
       st410_globaltree_plot, width = 12, height = 14)

ggsave(here("manuscript/esco/figures/FIG_5_st410_treeplot.pdf"),
       st410_globaltree_plot, width = 12, height = 14)

st410_globaltree_plot

```



```{r st410 subtree, fig.width = 7, fig.height = 6}
treeio::tree_subset (st410_tree, 564, levels_back = 0) -> st410_subtree

esbl_cpe.ariba410 %>%  
  subset(sample %in% st410_subtree$tip.label) %>% 
  as.data.frame() ->
  esbl_cpe.ariba410_subset
  
  esbl_cpe.ariba410_subset[
    c(TRUE, apply(esbl_cpe.ariba410_subset[-1], 2,function(x) !all(is.na(x))))] ->
    esbl_cpe.ariba410_subset

  (  
  (ggtree(treeio::tree_subset (st410_tree, 564, levels_back = 0)) %<+% (select(st410_metadata, accession, Country) %>% mutate(Country = if_else(Country == "Malawi", "Malawi",NA_character_)))  %>%
     gheatmap(
       select(st410_metadata, `Year`),
       font.size = 4,
       width = 0.05,
       colnames_position = "top",
       color = NA,
       colnames_offset_y = 5,
       colnames_angle = 90,
       hjust = 0
     )  + 
     scale_fill_viridis(name = "Year", limits = c(1975,2020)) +
     new_scale_fill()) %>% 
    gheatmap(
      select(st410_metadata, Continent),
      font.size = 4,
      width = 0.05,
      colnames_position = "top",
      color = NA,
      colnames_offset_y = 5,
      colnames_angle = 90,
      offset = 0.00003,
      hjust = 0
    ) + 
    scale_fill_viridis_d(option = "plasma", name = "Continent") + 
      geom_point2(aes(subset = as.numeric(label) < 95 & !isTip), size = 1) +
    geom_tippoint(aes(color = Country), na.rm = TRUE, size = 0.5) + 
    scale_color_manual(values  = "red", na.translate = FALSE) +
    new_scale_fill() 
) %>% 
  gheatmap(
    select(esbl_cpe.ariba410_subset, -sample),
    font.size = 3,
    width = 0.5,
    colnames_position = "top",
    color = NA,
    colnames_offset_y = 5,
    colnames_angle = 90,
    offset = 0.00006,
    hjust = 0
  ) + scale_fill_manual(values = viridis_pal(option = "magma")(5)[c(2,3)],
                        na.translate = FALSE,
                        name = "Beta-\nlactamase") +
  ylim(NA, 200) +
  geom_treescale(x = 0.0002, y = 250, offset = 2) +
    geom_highlight(aes(node = 181) , alpha = 0.3) +
    geom_cladelabel(
      node = 181,
      label = "B4/H24RxC",
      angle = 90,
      offset = -0.00032,
      barsize = 0,
      hjust = 0.5,
      fontsize = 4,
      offset.text = -0.00002,
      color = "steelblue"
    ) -> st410_subtree_plot
  
  ggsave(here("manuscript/esco/figures/FIG_6_st410_treeplot.svg"),
       st410_subtree_plot, width = 7, height = 6)

  ggsave(here("manuscript/esco/figures/FIG_6_st410_treeplot.pdf"),
       st410_subtree_plot, width = 7, height = 6)
  
st410_subtree_plot
  
```

### ST410 - mapping to plasmid mapped back to phylogeny

```{r plasmid_mapping, fig.width=7, fig.height=6}

 
bind_rows(
  read_csv(here("data_processed/esco/st167/plasmid_mapping/st167_plam_cov.csv")) %>% 
    mutate(ST = "ST167"),
  read_csv(here("data_processed/esco/st410/plasmid_mapping/st410_plam_cov.csv")) %>% 
    mutate(ST = "ST410")
  ) %>% 
  mutate(file = gsub("\\./","", file),
         file = gsub("_filtered","", file),
         file = gsub("/snps\\.aligned\\.fa","", file),
         file = gsub("#", "_", file),
         coverage = mapped_bases/ref_len,
         identity = (mapped_bases-snp_dist_to_ref)/mapped_bases) %>% 
  
  as.data.frame() ->
  plasmid_mapping

rownames(plasmid_mapping) <- plasmid_mapping$file



(
  (ggtree(
    treeio::tree_subset (st410_tree, 1083, levels_back = 0)
  )  %<+% (
    select(st410_metadata, accession, Country) %>% mutate(Country = if_else(
      Country == "Malawi", "Malawi", NA_character_
    ))
  )) %>%
    gheatmap(
      select(esbl_cpe.ariba410, c("NDM-5", "OXA-181")),
      font.size = 2.5,
      width = 0.1,
      colnames_position = "top",
      color = NA,
      colnames_offset_y = 2,
      colnames_angle = 90,
      offset = 0,
      hjust = 0
    ) + ylim(NA, 580) +
    geom_tippoint(aes(color = Country), na.rm = TRUE, size = 0.5) +
    scale_color_manual(
      values  = viridis_pal(option = "magma")(4)[3],
      na.translate = FALSE
    )  + scale_fill_manual(
      values = viridis_pal(option = "magma")(5)[c(2, 3)],
      na.translate = FALSE,
      name = "Beta-\nlactamase",
      guide = FALSE
    ) +
    new_scale_fill()
) %>%
  facet_plot(panel = 'Coverage', data = plasmid_mapping,
             geom = geom_barh,
             mapping = aes(x = coverage, fill = "meh"),
             stat='identity' ) %>%
  facet_plot(panel = 'Identity', data = plasmid_mapping,
             geom = geom_barh,
             mapping = aes(x = identity, fill = "meh"),
             stat='identity' ) +
  scale_fill_manual(values = viridis_pal(option = "cividis")(7)[2]) +
  theme_classic() + 
  theme(legend.position = "none") ->
  st410_plasmid_mapping_plot

st410_plasmid_mapping_plot
ggsave(here("manuscript/esco/figures/SUPP_FIG_3_st410_plasm_mapping_plot.svg"),
       st410_plasmid_mapping_plot, width = 7, height = 6)

(
  (ggtree(
    treeio::tree_subset (st167_tree, 320, levels_back = 0)
  )  %<+% (
    select(st167_metadata, accession, Country) %>% mutate(Country = if_else(
      Country == "Malawi", "Malawi", NA_character_
    ))
  )) %>%
    gheatmap(
      select(esbl_cpe.ariba, c("NDM-5", "OXA-181")),
      font.size = 2.5,
      width = 0.1,
      colnames_position = "top",
      color = NA,
      colnames_offset_y = 2,
      colnames_angle = 90,
      offset = 0,
      hjust = 0
    ) + ylim(NA, 340) +
    geom_tippoint(aes(color = Country), na.rm = TRUE, size = 0.5) +
    scale_color_manual(
      values  = viridis_pal(option = "magma")(4)[3],
      na.translate = FALSE
    )  + scale_fill_manual(
      values = viridis_pal(option = "magma")(5)[c(2, 3)],
      na.translate = FALSE,
      name = "Beta-\nlactamase",
      guide = FALSE
    ) +
    new_scale_fill()
) %>%
  facet_plot(panel = 'Coverage', data = plasmid_mapping,
             geom = geom_barh,
             mapping = aes(x = coverage, fill = "meh"),
             stat='identity' ) %>%
  facet_plot(panel = 'Identity', data = plasmid_mapping,
             geom = geom_barh,
             mapping = aes(x = identity, fill = "meh"),
             stat='identity' ) +
  scale_fill_manual(values = viridis_pal(option = "cividis")(7)[2]) +
  theme_classic() + 
  theme(legend.position = "none") ->
  st167_plasmid_mapping_plot




```

### ST167 - mapping to plasmid mapped back to phylogeny

```{r st167-plasm, fig.width=7, fig.height=6}

st167_plasmid_mapping_plot
ggsave(here("manuscript/esco/figures/SUPP_FIG_2_st167_plasm_mapping_plot.svg"),
       st167_plasmid_mapping_plot, width = 7, height = 6)



```