---
title: "Figures and tables: _E. coli_ genomics"
author: "Joe Lewis"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggtree)
library(ape)
library(phytools)
library(tidyverse)
library(blantyreESBL)
library(lubridate)
library(ggnewscale)
library(countrycode)
library(viridis)
library(blantyreESBL)
library(kableExtra)
library(patchwork)
library(here)
library(ggnewscale)
library(scales)
```

## Introduction

This document generates figures and tables for Joe's _E. coli_ genomics paper.

### Assembly stats and accession numbers

```{r accession-table}

btESBL_sequence_sample_metadata %>%
  filter(lane %in% btESBL_coregene_tree_esco$tip.label) %>%
  left_join(
    rbind(
      read_tsv(
        "~/Documents/PhD/Thesis/bookdown/chapter_7/checkm_quast/D1/transposed_report.tsv"
      ) ,
      read_tsv(
        "~/Documents/PhD/Thesis/bookdown/chapter_7/checkm_quast/D220190318/transposed_report.tsv"
      ),
      read_tsv(
        "~/Documents/PhD/Thesis/bookdown/chapter_7/checkm_quast/D220190503/transposed_report.tsv"
      )
    ) %>%
      transmute(
        lane = gsub("\\.contigs_spades", "",  Assembly),
        number_of_contigs = `# contigs`,
        N50 = N50
      ),
    by = "lane"
  ) %>% 
  select(accession, lane, supplier_name, pid, number_of_contigs, N50) -> t1

t1 %>% 
  kbl(caption = "Accession numbers and assembly statistics for included samples") %>% 
  kable_classic(full_width = FALSE)

```

### Phylogroup and MLST

```{r mlst-and-phylogroup, fig.height = 4, fig.width = 7, fig.cap = "FIGURE 1: Sequency types (A) and phylogroups (B) of included isolates"}

left_join(
  read_csv("~/Documents/PhD/Thesis/bookdown/chapter_7/phylogroup_and_mlst/mlst.csv"),
  read_csv("~/Documents/PhD/Thesis/bookdown/chapter_7/phylogroup_and_mlst/phylogroups.csv"),
  by = c("lane" = "Lane")
  ) -> 
  pgroup_mlst

pgroup_mlst %>% 
  group_by(ST) %>% 
  mutate(n = n()) %>% 
  filter(n > 2) %>% 
  ggplot(aes(fct_rev(fct_infreq(ST)), fill = Phylogroup)) +
  geom_bar() +
  coord_flip() +
  theme_bw() +
  theme(legend.position = "none") +
  labs(x = "ST", y = "n") -> p1 
#  scale_fill_viridis_d(option = "cividis") -> p1
  
pgroup_mlst %>% 
  ggplot(aes(fct_infreq(Phylogroup), fill = Phylogroup)) +
  geom_bar() +
  theme_bw() +
  theme(legend.position = "none", 
        axis.text.x = element_text(angle = 45,
                                   hjust = 1)) +
#  scale_fill_viridis_d(option = "cividis") +
  labs(x = "Phylogroup", y = "n") -> p2
  
(p1 | p2) + plot_annotation(tag_levels = "A") -> mlst_plot

ggsave(here("manuscript/esco/figures/FIG_1_mlstplot.svg"), mlst_plot,
       width = 7, height = 4)

mlst_plot 
```

### Malawi phylogeny

```{r malawi-tree, fig.height=9, fig.width= 9, fig.cap = "SUPPLEMENTARY FIGURE 1: Midpoint-rooted maximum-likelohood phylogeny of study isolates shwoing phylogroup and multilocus sequence type."}

pgroup_mlst %>% 
  as.data.frame() -> pgroup_mlst

rownames(pgroup_mlst) <- pgroup_mlst$lane

pgroup_mlst %>% 
  select(lane, ST) %>% 
  group_by(ST) %>% 
  mutate(n = n(),
         ST = if_else(n ==1, "Other", ST)) %>% 
#  filter(ST != "Novel") %>% 
  mutate(ST = if_else(!ST %in% c("Novel", "Other"), 
                      paste0("ST", ST), ST)) %>% 
  arrange(fct_infreq(ST)) %>% 
  pivot_wider(id_cols = lane, 
              names_from = ST,
              values_from = ST,
              values_fn = length,
              values_fill = 0) %>% 
  mutate(across(everything(), as.character)
         ) %>% 
  relocate(Other, .after = everything()) %>% 
  as.data.frame() ->
  mlst_onehot

rownames(mlst_onehot) <- mlst_onehot$lane

(
  ggtree(btESBL_coregene_tree_esco) %>%
    gheatmap(
      select(pgroup_mlst, Phylogroup),
      width = 0.1,
      color = NA,
      font.size = 4,
      colnames_angle = 90,
      colnames_position = "top",
      colnames_offset_y = 3,
      hjust = 0
    ) +
    scale_fill_manual(
      values = hue_pal()(8),
      name = "Phylogroup") + 
   new_scale_fill()
) %>%
  gheatmap(select(mlst_onehot, -lane),
           font.size = 2.5,
           color = NA,
           colnames_angle = 90,
           offset = .03,
           colnames_offset_y = 3,
           colnames_position = "top",
           hjust = 0)  +
  scale_fill_manual(values = c("lightgrey", "black"), guide = FALSE) +
  ylim(NA, 560)  + 
  annotate("text", x= 0.28, y =520, label = "Sequence Type") -> malawi_treeplot

ggsave(here("manuscript/esco/figures/SUPP_FIG_1_malawi_treeplot.svg"), malawi_treeplot, width = 9, height = 9)


```

### AMR determinents

```{r amrplot, fig.width= 6, fig.height=10}

read_csv("~/Documents/PhD/Thesis/bookdown/chapter_7/amr/srst2_pres_abs.csv") %>% filter(name %in% btESBL_coregene_tree_esco$tip.label) -> amr.ariba
  

amr.qrdr <- read_csv( "~/Documents/PhD/Thesis/bookdown/chapter_7/amr/qrdr_mut_presence.csv")

# get beta lactamase phenotype and tidy -------------------------------------

bls <- read_tsv(
  "https://ftp.ncbi.nlm.nih.gov/pathogen/betalactamases/Allele.tab"
  )


bls %>% 
  rename_with(~ tolower(gsub(" ", "_", .x))) %>% 
  rename_with(~ gsub("#", "", .x)) %>% 
  mutate(allele_name = gsub("-", "_", allele_name),
         class = case_when(
           grepl("carbapenem-hydrolyzing", 
                 curated_gene_product_name) ~ "Carbapenemase",
           grepl("metallo-beta-lactamase", 
                 curated_gene_product_name) ~ "Carbapenemase",
           grepl("extended-spectrum beta-lactamase", 
                 curated_gene_product_name) ~ "ESBL",
           grepl("class C", 
                 curated_gene_product_name) ~ "AmpC",
           grepl("beta-lactamase", 
                 curated_gene_product_name) ~ "Penicillinase"
         )) -> 
  bls




# add class ro which resistnce is conferred --------------------------------

quinolone <- "Par|Gyr|Par|Qnr|Qep|Nor|GyrA|GyrB|ParC|ParE"
tetracycline<- "Tet"
sulphonamide <- "Sul"
aminoglycoside <- "Str|Aad|Aac|Aph|Rmt|APH"
streptothricin <- "Sat"
macrolide <- "Mph|Mdf|Erm|Ere"
fosfomycin <- "Fos"
chloramphenicol <- "Cat|FloR|Cml"
trimethoprim <- "Dfr"
rifampicin <- "Arr"
ESBL <- "SHV_12"
penicillinase <- "OKP|SCO|LEN|LAP"
ampc <- "CMY"

amr.ariba %>% 
  pivot_longer(-name, names_to = "gene") %>% 
  bind_rows(
    amr.qrdr %>% 
    pivot_longer(-sample, names_to = "gene") %>% 
      rename(name = sample)
    ) %>% 
  filter(gene != "AmpH") %>% 
  left_join(select(bls, allele_name, class),
            by = c("gene" = "allele_name")) %>% 
  mutate(class = case_when(
    str_detect(gene, quinolone) ~ "Quinolone",
    str_detect(gene, tetracycline) ~ "Tetracycline",
    str_detect(gene, sulphonamide) ~ "Sulphonamide",
    str_detect(gene, aminoglycoside) ~ "Aminoglycoside",
    str_detect(gene, streptothricin) ~ "Streptothricin",
    str_detect(gene, macrolide) ~ "Macrolide",
    str_detect(gene, fosfomycin) ~ "Fosfomycin",
    str_detect(gene, chloramphenicol) ~ "Chloramphenicol",
    str_detect(gene, rifampicin) ~ "Rifampicin",
    str_detect(gene,trimethoprim) ~ "Trimethoprim",
    str_detect(gene,ESBL) ~ "ESBL",
    str_detect(gene,penicillinase) ~ "Penicillinase",
    str_detect(gene,ampc) ~ "AmpC",
    TRUE ~ class
  )) ->
  amr.ariba.long



## AMR

### Overall prevalence


amr.ariba.long %>% 
  filter(value == 1) %>% 
  filter(!is.na(class)) %>% 
  group_by(class) %>% 
  mutate(n_class = length(class),
         n_genes_in_class = n_distinct(gene)) %>% 
  select(class, n_class,n_genes_in_class) %>% 
  unique() %>% 
  arrange(n_class) %>% 
  ungroup() %>% 
  mutate(end = cumsum(n_genes_in_class),
         start = lag(end, default = 0),
         textpos = start+0.6 + 0.5*(end-start)) -> annotate.df

amr.ariba.long %>% 
  filter(value == 1) %>% 
  filter(!is.na(class)) %>% 
  group_by(class) %>% 
  mutate(n_class = length(class),
         n_genes_in_class = n_distinct(gene),
         gene = gsub("_", "-", gene)) %>% 
ggplot( aes(fct_reorder(fct_rev(fct_infreq(gene)), n_class), 
            fill = class)) + 
  geom_bar() +
  theme_bw() +
  coord_flip(ylim = c(-5,450),clip = "off", expand = FALSE, xlim = c(0, 76)) +
  annotate(geom = "segment",
             x = annotate.df$start +0.5+0.2, 
             xend = annotate.df$end +0.5-0.2, 
             y = -115, yend = -115 ) +
  annotate(geom = "text", y = -125,
           x = annotate.df$textpos,
           label = annotate.df$class, 
           size = 3, 
           hjust = 1) +
  labs(y = "Number") + 
  theme(plot.margin = unit(c(0.2,0.5,0.2,4), "cm"),
        axis.title.y  = element_blank(),
        legend.position = "none") -> amrplot

amrplot
ggsave(filename = here("manuscript/esco/figures/FIG_2_amr_plot.svg"),plot = amrplot,
       width =6,
       height = 10
)

```



